File: app/src/main/java/com/xianzhitech/ptt/model/Privilege.java
package com.xianzhitech.ptt.model;

import android.support.annotation.IntDef;

import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;

/**
 * Created by fanchao on 15/12/15.
 */
@Retention(RetentionPolicy.SOURCE)
@IntDef(flag = true, value = { Privilege.MAKE_CALL, Privilege.CREATE_GROUP, Privilege.RECEIVE_CALL, Privilege.RECEIVE_GROUP })
public @interface Privilege {
    int MAKE_CALL = 1;
    int CREATE_GROUP = 1 << 1;
    int RECEIVE_CALL = 1 << 2;
    int RECEIVE_GROUP = 1 << 3;
}

File: app/src/main/kotlin/com/xianzhitech/ptt/model/Privilege.kt
package com.xianzhitech.ptt.model

import java.util.*

enum class Privilege {
    MAKE_CALL,
    CREATE_ROOM,
    RECEIVE_CALL,
    RECEIVE_ROOM;
}

fun String?.toPrivileges(): EnumSet<Privilege> {
    val result = EnumSet.noneOf(Privilege::class.java)
    this?.foldIndexed(0, { index, lastIndex, c ->
        if (c == '|') {
            result += Privilege.valueOf(substring(lastIndex, index))
            index + 1
        }
        else {
            lastIndex
        }
    })
    return result
}

fun Iterable<Privilege>?.toDatabaseString() = this?.joinToString(separator = "|")